symengine/parser.cpp

```diff
@@ -0,0 +1,326 @@
+                // using the recursive call to parse_string
+                if (s[iter] == '+') {
+                    result = add(result, parse_string(iter+1, operator_end[iter]));
+                    iter = operator_end[iter]-1;
+
+                } else if (s[iter] == '*') {
+                    result = mul(result, parse_string(iter+1, operator_end[iter]));
+                    iter = operator_end[iter]-1;
+
+                } else if (s[iter] == '-') {
+                    result = sub(result, parse_string(iter+1, operator_end[iter]));
+                    iter = operator_end[iter]-1;
+
+                } else if (s[iter] == '/') {
+                    result = div(result, parse_string(iter+1, operator_end[iter]));
+                    iter = operator_end[iter]-1;
+
+                } else if (s[iter] == '^') {
+                    result = pow(result, parse_string(iter+1, operator_end[iter]));
+                    iter = operator_end[iter]-1;
+
+                } else if (s[iter] == '(') {
+                    result = functionify(iter, expr);
+
+                } else {
+                    continue;
+                }
+                // result has definitely been set
+                result_set = true;
+
+            } else {
+                // if not an operator, we append it to the current expr
+                expr += s[iter];
+                // check wether it's numeric or not
+                if (s[iter] == '.') {
+                    num_dots++;
+
+                } else {
+                    int ascii = s[iter] - '0';
+                    if (ascii < 0 or ascii > 9)
+                        is_not_numeric = true;
+                }
+                // if the parsing was to finish after this, result must be set
+                // occurs when no operator present eg. "3"
+                if (iter == h-1)
+                    result = set_result(expr, is_not_numeric, num_dots);
+            }
+        }
+
+        return result;
+    }
+
+    // returns true if the s[iter] is an operator
+    bool is_operator(int iter)
+    {
+        if (iter >= 0 and iter < (int)s_len)
+            if (OPERATORS.find(s[iter]) != OPERATORS.end())
+                return true;
+        return false;
+    }
+
+    // is called on detecting a "func(", thus "func" must be a type of
+    // function and everything inside arguments of the function
+    RCP<const Basic> functionify(unsigned int& iter, const std::string& expr)
+    {
+        vec_basic params;
+
+        while (s[iter] != ')') {
+            params.push_back(parse_string(iter+1, operator_end[iter]));
+            iter = operator_end[iter];
+        }
+
+        // make this modular, scalable
+        // make map for all those functions which take in vec_basic args, like levi_civita
+        if (params.size() == 1)
+            if (single_arg_functions.find(expr) != single_arg_functions.end())
+                return single_arg_functions[expr](params[0]);
```

Comment:
We can do this later once we have some benchmarks, but in general, I think you want to save the result of the `.find()` method and then use it to access the element instead of `single_arg_functions[expr]`, because as it is now, there are two search for the `expr` element in the `single_arg_functions`.